Support for server side apply

redbaron opened the pull request:
First take on server side apply feature  (https://github.com/vmware-tanzu/carvel-kapp/issues/388) to collect early feedback.

Some highlights:
- user provided manifest + some kapp generated annotations is what we are submitting to K8S via PATCH request. 
- diff calculation is at the core of `kapp` code paths, so I had to approximate it by diffing existing instances against dry run SSA result. This required calls to K8S much earlier in the code than it was previously, which required some plumbing
 - because we have to make dry run call, there is a new  `NewChangeSSA` change factory
 - E2E test runner now supports `KAPP_E2E_SSA=1` env var value. When set, `--ssa-enabled` flag is added to kapp invocation allowing same tests to be run in SSA mode. Tests with rebasing are skipped in SSA mode.
 - To make debugging e2e cases easier I created embeddable runner, which runs kapps `*corbra.Command` in the testing process. It is not a full replacement for a `kapp` command execution, but greatly helped developing this feature as most interesting scenarios come up in `e2e` and not   unit testing.

Unimplemented:
- [x] SSA conflict force flag. It is present, but not used anywhere.
- [x] conflicting CLI flags are ignored in favour of SSA feature. I think there should be  a clear error message if conflicting flags are specificed, like `--diff-against-last-applied`.

Yet untested (unless there is an existing E2E test covering it):
- [ ] how changes to namespace are handled
- [ ] how changes to apiVersion are handled (apps/v1beta1 -> apps/v1)

Features deliberately left outside of this PR:
- Ownership transfer as described in [docs](https://kubernetes.io/docs/reference/using-api/server-side-apply/#transferring-ownership).  There might be situations where user might want to give up ownership of certain fields after initial creation.  `kapp` could assist with it.
- Fine grained force conflict resolution. Right now it is all or nothing. There is no fine grained force option in the kubernetes, so it has to be implemented as 2 separate operations: non-SSA patch for conflicting fields + SSA for the rest. I'll create an issue with more detailed proposal once SSA work is merged

redbaron committed:
Adds e2e test for simple server side apply merge case

redbaron committed:
First basic SSA support

redbaron committed:
Embeddable E2E tests

100mik commented:
Thanks for the draft PR @redbaron !
Will definitely go through the changes to get an idea of what's going on.

redbaron committed:
Fix tests compilation

redbaron committed:
Make linter happy

redbaron committed:
Fix diffing logic so that it never diffs managed fields

redbaron commented on pkg/kapp/diff/change_factory.go:
```diff
@@ -4,26 +4,76 @@
 	if existingRes != nil {
+		// Strip rebasing "base" object of kapp history so that it never shows up in the diff, regardless
+		// of rebase rules used
+		existingResForRebasing, err = f.NewResourceWithHistory(existingRes).HistorylessResource()
```
I believe there was a sleeper bug here, which unfolded like following:
- By the end of this function `existingRes` was unconditionally stripped of history with `HistorylessResource()`
- `newRes` was unconditionally stripped of history with `HistorylessResource()`
- `existingResForRebasing` has history kept, because it comes from the cluster as is
- When building `rebasedNewRes` using `existingResForRebasing + newRes`, history was removed due to default rebase rules included in the kapp transplanting history-less annotations from `newRes` 
- If we redeploy same object as previous deployment, end result has no spurious diff, because both objects have history stripped

Should provided config had different rebase rules, `rebasedNewRes` is going to have history annotations from `existingResForRebasing`, but `existingRes` wont, generating spurious diff.

Similar situation arises when `metadata.managedFields` is populated by the cluster. This field is considered history and is stripped in `HistorylessResource()` call, but default rebase rules have no mentions of it.

Instead of updating default rebase rule, I opted for fixing it at the source. Here I am stripping history from all objects, ensuring it is not going to show up in any diffs regardless of kapp config provided. Technically it is a change in behaviour, hence this comment to highlight it.

cppforlife replied:
will think through it, thanks for call out 👍

cppforlife replied:
to double check -- changes in NewChangeAgainstLastApplied are not really affecting SSA, right? (i might have to ask to separate them out into a diff PR).

cppforlife commented on pkg/kapp/cmd/app/apply_flags.go:
```diff
@@ -62,6 +64,10 @@ func (s *ApplyFlags) SetWithDefaults(prefix string, defaults ApplyFlags, cmd *co
 	cmd.Flags().IntVar(&s.WaitingChangesOpts.Concurrency, prefix+"wait-concurrency",
 		5, "Maximum number of concurrent wait operations")
 
+	cmd.Flags().BoolVar(&s.ServerSideApply, prefix+"server-side", false, "If true, apply runs in the server instead of the client")
+	cmd.Flags().StringVar(&s.FieldManagerName, prefix+"field-manager", "kapp-server-side-apply", "Name of the manager used to track field ownership")
+	cmd.Flags().BoolVar(&s.ServerSideForceConflict, prefix+"force-conflicts", false, "If true, server-side apply will force the changes against conflicts.")
```
lets prefix these flags with server-side-apply (or may be just ssa) e.g.

--server-side-apply
--server-side-apply-field-manager
--server-side-apply-force-conflicts

or 

--ssa
--ssa-field-manager
--ssa-force-conflicts

cppforlife replied:
also are there cases when somebody would want to force conflicts for one but not the other resource? what would be workflow there, or is it pretty much always --ssa-force-conflicts=true (and if so should that be default?)

redbaron replied:
I kept flag names and defaults as they are in `kubectl apply`, which many people are familiar with.

redbaron replied:
Done

cppforlife commented on pkg/kapp/cmd/app/deploy.go:
```diff
@@ -51,7 +52,12 @@ func NewDeployCmd(o *DeployOptions, flagsFactory cmdcore.FlagsFactory) *cobra.Co
 		Short:   "Deploy app",
-		RunE:    func(_ *cobra.Command, _ []string) error { return o.Run() },
+		RunE: func(_ *cobra.Command, _ []string) error {
+			if o.ApplyFlags.ServerSideApply {
+				o.DiffFlags.ChangeSetOpts.Mode = ctldiff.ServerSideApplyChangeSetMode
+			}
```
not a fan of doing it in RunE, lets move this into Run()... (also going to see if there is a better place for this)

redbaron replied:
I plan to add flag validation step and return error, rather than silently overwriting values provided by user, so it should be a RunE I believe because it can return error.

100mik replied:
I believe we could move these validations to the `Run()`function  as it returns an error and let the command initialisation have
```go
RunE:    func(_ *cobra.Command, _ []string) error { return o.Run() }
```

redbaron replied:
Moved to PreRunE

cppforlife commented on pkg/kapp/cmd/app/deploy.go:
```diff
@@ -315,19 +324,17 @@ func (o *DeployOptions) existingResources(newResources []ctlres.Resource,
-func (o *DeployOptions) calculateAndPresentChanges(existingResources,
-	newResources []ctlres.Resource, conf ctlconf.Conf, supportObjs FactorySupportObjs) (
-	ctlcap.ClusterChangeSet, *ctldgraph.ChangeGraph, bool, string, error) {
+func (o *DeployOptions) calculateAndPresentChanges(ctx context.Context, existingResources, newResources []ctlres.Resource, conf ctlconf.Conf, supportObjs FactorySupportObjs) (ctlcap.ClusterChangeSet, *ctldgraph.ChangeGraph, bool, string, error) {
```
nit: undo formatting change -- easier to see when it's on multiple lines

redbaron replied:
done

cppforlife commented on pkg/kapp/cmd/tools/diff_flags.go:
```diff
@@ -35,7 +35,11 @@ func (s *DiffFlags) SetWithPrefix(prefix string, cmd *cobra.Command) {
 	cmd.Flags().BoolVar(&s.Mask, prefix+"mask", true, "Apply masking rules")
 
-	cmd.Flags().BoolVar(&s.AgainstLastApplied, prefix+"against-last-applied", true, "Show changes against last applied copy when possible")
+	if *cmd.Flags().Bool(prefix+"against-last-applied", true, "Show changes against last applied copy when possible. Ignored when server-side apply is used.") {
```
nit: lets put (`Ignored when server-side apply is used.` into parenthesis). we typically dont have multiple sentences in descriptions.

redbaron replied:
Done. Also changed wording to indicate conflict.

cppforlife commented on pkg/kapp/diff/change_factory.go:
```diff
@@ -4,26 +4,76 @@
+
+func (f ChangeFactory) NewChangeSSA(ctx context.Context, existingRes, newRes ctlres.Resource) (Change, error) {
+	if existingRes != nil {
+		historylessExistingRes, err := f.NewResourceWithHistory(existingRes).HistorylessResource()
```
should we even bother with "history" when doing SSA?

redbaron replied:
You are right. Because diff is between existing and dry run apply result, history remains unchanged between these two, so no need to strip it.

cppforlife commented on test/e2e/diff_test.go:
```diff
@@ -120,8 +120,9 @@ data:
-		out, _ := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name, "--json"},
+		out, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name, "--json"},
 			RunOpts{IntoNs: true, StdinReader: strings.NewReader(yaml1)})
+		require.NoError(t, err)
```
no need to check on err -- by default kapp object should do it for you unless AllowError is set to true. (this just makes it less noisy to write tests)

redbaron replied:
I'll do same error handling in RunEmbedded then

redbaron replied:
Done

cppforlife commented on pkg/kapp/diff/change_set.go:
```diff
@@ -23,13 +32,24 @@ type ChangeSet struct {
 func NewChangeSet(existingRs, newRs []ctlres.Resource,
 	opts ChangeSetOpts, changeFactory ChangeFactory) *ChangeSet {
 
+	if opts.Mode == "" {
+		opts.Mode = ExactChangeSetMode
+	}
```
seems like we should be able to make tighter assertion that mode isnt empty instead of just defaulting.

redbaron replied:
It is to preserve original behaviour where empty `ChangeSetOpts{}` used  exact change factory function.  Empty `ChangeSetOpts{}` is used in tests only.

cppforlife reviewed (commented):
looking pretty good so far.

praveenrewar commented on test/e2e/diff_test.go:
```diff
@@ -87,7 +87,7 @@ data:
 			"kind":            "ConfigMap",
 			"name":            "redis-config",
-			"namespace":       "kapp-test",
+			"namespace":       env.Namespace,
```
💯

redbaron committed:
Cosmetic changes to address feedback

redbaron committed:
Remove unused field

redbaron committed:
Don't strip history in ChangeFactory.NewChangeSSA because it won't show up in the diff anyway

100mik commented on pkg/kapp/resources/resources.go:
```diff
@@ -293,7 +302,7 @@ func (c *ResourcesImpl) Update(resource Resource) (Resource, error) {
 }
 
-func (c *ResourcesImpl) Patch(resource Resource, patchType types.PatchType, data []byte) (Resource, error) {
+func (c *ResourcesImpl) Patch(resource Resource, patchType types.PatchType, data []byte, dryRun bool) (Resource, error) {
```
I wonder if we  should add something like a struct `resource.PatchOpts` and pass that to the function instead.
Something like 
```go
type PatchOpts struct {
    dryRun bool
}
```
This allows us to add options in the future easily without changing the function definition. 🤔

redbaron replied:
Done

redbaron committed:
Another fix for NewChangeAgainstLastApplied to prevent history showing up in diff

redbaron committed:
Move dryRun to PatchOpts struct

redbaron committed:
Make RunEmbedded to fail tests on error

redbaron committed:
Adjust tests to reflect that history is not passed to rebasing rules

redbaron commented on test/e2e/config_test.go:
```diff
@@ -313,6 +313,8 @@ apiVersion: v1
 metadata:
   name: test-cm
+  annotations:
+    ann1: val1
```
I've added this to show that not all annotations stripped when object is passed to rebase rules

redbaron commented on test/e2e/config_test.go:
```diff
@@ -351,15 +353,18 @@ data:
 			"values": asYAML(t, map[string]interface{}{
 				"existing": func() interface{} {
 					raw := cm.Raw()
+					removeHistory(raw)
```
Consequence of https://github.com/vmware-tanzu/carvel-kapp/pull/392/files#r768970039  is that rebase rules don't see history in objects passed to them. I've adjusted test to reflect it.

redbaron commented:
Hm, all tests are passing locally on Docker Desktop managed 1.21 kubernetes cluster, but  `TestSSAUpdateSimple`  fails in CI.

redbaron committed:
Add SSA support for diff command. Check CLI flags for conflicts

redbaron committed:
Add copyright and license header

redbaron committed:
Run all e2e tests as kapp subprocess. Remove unnecessary error check.

redbaron committed:
Fix diff generation for versioned renamed objects

redbaron committed:
Support enabling SSA for E2E tests via env var globally

redbaron committed:
Fix panic in delete

100mik commented:
Looks like it is just the new test. Let me take a swing at it as well 🤔

cppforlife commented:
(will also circle back in beginning of Jan, happy holidays!)

redbaron commented:
> your company's legal contact did not review your signed contributor license agreement within the 14 day limit.

I don't understand, which company? Which address your bot contacted? I am contributing from my personal account.

praveenrewar commented:
> Hm, all tests are passing locally on Docker Desktop managed 1.21 kubernetes cluster, but TestSSAUpdateSimple fails in CI.

The tests are running on k8s v1.16 and v1.22, and I think the behaviour is a bit different on those.
For v1.22 you might wanna remove the `spec.selector` field in `kubectlChange` (I haven't looked at it in detail so I am not sure how that will effect the e2e test).

redbaron committed:
When SSA enabled, dry run object creation to get most accurate diff

redbaron committed:
Enable SSA during tests only if KAPPP_E2E_SSA is set to 1

redbaron committed:
Make IdentifierResources.Patch method to return Resource with stripped identity annotation

Other methods, like Create and Update, are stripping annotations,
this change makes Patch consistent with the rest.

Other methods force identity annotation to be set on all resources sent to K8S API.
Technically Patch should do the same with patch data, but it is tricky as there are multiple
patch formats supported by K8S and analysing and amending them to force identity annotation
to be present doesn't worth the effort.

redbaron committed:
Persist history using Patch instead of Update call.

This makes it unnecessary to have retry logic as no patch errors are retriable

redbaron committed:
FIXME: Temporarily remove Create dry run, because it wasn't dry run at all

redbaron committed:
Fixme IdentifierResources.Patch method

redbaron committed:
Fixme Persist history using Patch

redbaron committed:
Add support for SSA to 'Add*' strategies

redbaron committed:
Clear history when creating SSA change

redbaron committed:
Drop ChangeSSA as it becomes exactly like ChangeImpl. Use latter instead.

redbaron committed:
env.Namespace some of the tests

redbaron committed:
Skip diff comparison in template test.

SSA legitimately generates different diff

redbaron committed:
Carry on generating SSA change even if attempting to make an invalid update

Invalid updates will either be retried as replace, in which case generated diff is accurate
or it will be failed at the later stage after change graph is generated

redbaron committed:
env.Namespace more tests

redbaron committed:
Skip conflict rebasing test in SSA mode

redbaron committed:
env.Namespace more tests

redbaron committed:
Use JSON merge patch when recording history

Strategic merge patch is not support for CRDs without schema

redbaron committed:
Add SSASkip and comments

redbaron committed:
Run E2E tests in SSA and non-SSA mode

redbaron committed:
Fix nil panic in tests

redbaron committed:
Remove RunEmbedded leftovers in e2e tests

redbaron committed:
Use FieldManagerName from CLI args in the AddPlainStrategy

redbaron committed:
Add support for ssa-force flag

redbaron committed:
Make linter happy

redbaron committed:
Remove unused file

redbaron commented:
OK, it reached the stage where I'd like to propose it for a merge.

I added option to run E2E tests with SSA enabled and  test suite now passes with SSA on my 1.21 K8S,  except for tests with rebasing. I'll go through diff and add notes on  parts worth paying attention to.

`TestIgnoreFailingAPIServices` fails for me locally even on master, because error message text differs from what test expects.

There is a missing test for `--ssa-force` option, I'll add it in due course, but otherwise I consider initial implementation complete.
I'll also update  PR message to reflect current state shortly.

/cc @cppforlife

cppforlife commented:
cool cool. ill do a deep dive today/tomorrow on this @redbaron. other folks will also take a look...

cppforlife commented:
(still reading through the code... hoping to be done tomorrow)

cppforlife commented on pkg/kapp/cmd/app/deploy_flag_help_sections.go:
```diff
@@ -17,6 +17,10 @@ var (
 		PrefixMatch: "diff",
 	}
+	SSAFlagGroup = cobrautil.FlagHelpSection{
+		Title:       "Server side apply Flags:",
```
nit: "Server side apply Flags:" => "Server Side Apply Flags:"

redbaron replied:
done

cppforlife commented on test/e2e/update_server_side_test.go:
```diff
@@ -0,0 +1,130 @@
+
+	expectedObj := corev1.Service{}
+
+	// Patch dry run returns merged object with all patch-file fields present
+	err = kubectl.RunWithOptsIntoJSON([]string{"patch", "svc", "redis-primary", "--patch-file", tmpFile.Name(), "--dry-run=client"},
+		RunOpts{IntoNs: true}, &expectedObj)
+	require.NoError(t, err)
```
seems like a complicated way of determining expected state. is there something simpler that can be done (aka more static/hard-coded into the test)?

cppforlife commented on test/e2e/update_server_side_test.go:
```diff
@@ -0,0 +1,130 @@
+	require.NoError(t, err)
+
+	require.Equal(t, expectedObj, inClusterObj)
+
```
nit: remove new line

cppforlife commented on test/e2e/kapp.go:
```diff
@@ -96,10 +101,97 @@ func (k Kapp) RunWithOpts(args []string, opts RunOpts) (string, error) {
+		args = enableSSA(args)
+	}
+	var stdoutBuf bytes.Buffer
+	//var stdout io.Writer = bufio.NewWriter(&stdoutBuf)
```
nit: remove

cppforlife commented on test/e2e/kapp.go:
```diff
@@ -96,10 +101,97 @@ func (k Kapp) RunWithOpts(args []string, opts RunOpts) (string, error) {
+	return stdoutBuf.String(), err
+}
+
+func enableSSA(args []string) []string {
```
nit: place these methods on `func (Kapp)` to scope them to this class (nobody else should be using them).

cppforlife commented on pkg/kapp/cmd/tools/ssa/ssa_flags.go:
```diff
@@ -0,0 +1,30 @@
+}
+
+func (s *SSAFlags) Set(cmd *cobra.Command) {
+	cmd.Flags().BoolVar(&s.SSAEnable, "ssa-enable", false, "Use server side apply")
```
ssa-enable => ssa (this is in line with other flags like --wait=true|false).

can still keep struct field as SSAEnable if you want -- i would have went for something like since it's all under SSAFlags.

```
type SSAFlags struct {
	Enabled        bool
	ForceConflicts bool
	FieldManagerName string
}
```

redbaron replied:
done

cppforlife commented on pkg/kapp/resources/resources.go:
```diff
@@ -6,6 +6,7 @@ package resources
 import (
 	"context"
 	"fmt"
+	"github.com/k14s/kapp/pkg/kapp/cmd/tools/ssa"
```
resources pkg should not be importing things from cmd/* (commands depend on lower level packages, im guessing that's why you had to add ssa directory in tools.).

instead we should define either our own SSAOpts or have those fields on ResourcesImplOpts struct. SSAFlags struct in cmd/* can either embed this struct or just dup the fields.

redbaron replied:
done by duping fields

cppforlife commented on pkg/kapp/resources/resources.go:
```diff
@@ -305,9 +319,18 @@ func (c *ResourcesImpl) Patch(resource Resource, patchType types.PatchType, data
-		patchedUn, err = resClient.Patch(context.TODO(), resource.Name(), patchType, data, metav1.PatchOptions{})
+		patchedUn, err = resClient.Patch(context.TODO(), resource.Name(), patchType, data, metav1.PatchOptions{
+			FieldManager: c.opts.SSAFlags.FieldManagerName,
+			Force:        c.opts.SSAFlags.ForceParamValue(patchType),
```
lets localize forceParamValue logic to this class since it's only used here.

redbaron replied:
done.  I didn't test if Force flag on non Apply patch types breaks anything, so doing a safest option and passing it only on apply . It made logic a bit more convoluted, because we can't just pass `false` and every other case and has to pass `nil` instead.

cppforlife commented on pkg/kapp/cmd/app/factory.go:
```diff
@@ -56,16 +59,17 @@ func FactoryClients(depsFactory cmdcore.DepsFactory, nsFlags cmdcore.NamespaceFl
 		CoreClient:          coreClient,
 		ResourceTypes:       resTypes,
 		IdentifiedResources: identifiedResources,
+		Resources:           resources,
```
looks like this line is dead code

cppforlife commented on pkg/kapp/cmd/app/factory.go:
```diff
@@ -15,11 +16,12 @@ type FactorySupportObjs struct {
 	CoreClient          kubernetes.Interface
 	ResourceTypes       *ctlres.ResourceTypesImpl
 	IdentifiedResources ctlres.IdentifiedResources
+	Resources           ctlres.Resources
```
looks like this line is dead code

cppforlife commented on pkg/kapp/clusterapply/delete_change.go:
```diff
@@ -107,6 +107,6 @@ func (c DeleteOrphanStrategy) Apply() error {
 	}
 
-	_, err = c.d.identifiedResources.Patch(c.res, types.JSONPatchType, patchJSON)
+	_, err = c.d.identifiedResources.Patch(c.res, types.JSONPatchType, patchJSON, ctlres.PatchOpts{DryRun: false})
```
no need to specify `DryRun: false`

cppforlife commented on pkg/kapp/clusterapply/add_or_update_change.go:
```diff
@@ -218,36 +222,14 @@ func (c AddOrUpdateChange) recordAppliedResource(savedRes ctlres.Resource) error
 
-		return true, nil
-	})
+	_, err = c.identifiedResources.Patch(savedRes, types.MergePatchType, recordHistoryPatch, ctlres.PatchOpts{DryRun: false})
```
i guess this is safe-ish to do because metadata.annotations will always be present because kapp adds identity annotation.

(it's worth noting that "patch" verb is required in permissions in addition to update.)

cppforlife replied:
`DryRun: false` is not necessary

redbaron replied:
Even if `metadata.annotations` is not present MergePatch will create it. It would be a problem if we used JsonPatch.

cppforlife commented on pkg/kapp/clusterapply/add_or_update_change.go:
```diff
@@ -263,6 +245,26 @@ func (c AddPlainStrategy) Apply() error {
+
+	}
+
 	return c.aou.recordAppliedResource(createdRes)
```
there is a gap of time between create succeeding and patch succeeding where other changes may be applied to the resource. recordAppliedResource on the result of patch which means that we would capture such changes as "correlated" to initial submission. im not sure if this matters at all when ssa is enabled, but definitely seems suspicious if we keep it. at the least this deserves a comment, explaining why its ok for ssa.

cppforlife commented on pkg/kapp/clusterapply/add_or_update_change.go:
```diff
@@ -263,6 +245,26 @@ func (c AddPlainStrategy) Apply() error {
+`), ctlres.PatchOpts{DryRun: false})
+		if err != nil {
+			// TODO: potentially patch can fail if '"op": "test"' fails, which can happen if another
+			// controller changes managedFields. We
```
comment got cut off...

cppforlife commented on pkg/kapp/clusterapply/add_or_update_change.go:
```diff
@@ -295,7 +311,17 @@ type UpdatePlainStrategy struct {
+
+	if c.aou.opts.ServerSideApply {
+		updatedRes, err = ctlres.WithIdentityAnnotation(c.newRes, func(r ctlres.Resource) (ctlres.Resource, error) {
+			resBytes, _ := r.AsYAMLBytes()
```
lets throw in err check. its unlikely to kick in but just for safety.

cppforlife commented on pkg/kapp/clusterapply/add_or_update_change.go:
```diff
@@ -295,7 +311,17 @@ type UpdatePlainStrategy struct {
+			return c.aou.identifiedResources.Patch(r, types.ApplyPatchType, resBytes, ctlres.PatchOpts{DryRun: false})
+		})
+	} else {
+		updatedRes, err = c.aou.identifiedResources.Update(c.newRes)
```
should we be tryToResolveUpdateConflict doing when SSA is on?

redbaron replied:
You are right. `tryToResolveUpdateConflict` recalculates Change from a new existing resource. In case of SSA only conflict we can have is when we are conflicting with another fieldManager, no amount of new diffing going to help here.

cppforlife commented on pkg/kapp/clusterapply/add_or_update_change.go:
```diff
@@ -295,7 +311,17 @@ type UpdatePlainStrategy struct {
+	var err error
+
+	if c.aou.opts.ServerSideApply {
+		updatedRes, err = ctlres.WithIdentityAnnotation(c.newRes, func(r ctlres.Resource) (ctlres.Resource, error) {
```
it seems that we have two types of Patch operations: one where we are working with a resource, and one where we are working with some "minimal/raw" patch. instead of exposing WithIdentityAnnotation outside of IdentifiedResources resources class, lets keep existing Patch method to work with Resource class (instead of []byte) and change it so that it can add annotation on the fly. lets also add PatchRaw which does take []byte. this should result in keeping WithIdentityAnnotation as private.

cppforlife commented on pkg/kapp/diff/change_set.go:
```diff
@@ -4,14 +4,23 @@
+)
+
 type ChangeSetOpts struct {
-	AgainstLastApplied bool
+	Mode ChangeSetMode
```
after reading through flag "adjusting" above, i think we should just turns this into:

```
type ChangeSetOpts struct {
        ServerSideApply bool
	AgainstLastApplied bool
}
```

anywhere where we are doing a switch based on mode, we should just do:

```
switch {
case opts.ServerSideApply: ...
case opts.AgainstLastApplied: ...
default: ...
}
```

the reasoning for this, it seems that mode enum introduces quite a bit of gymanistics necessary to compute what it should be.

cppforlife commented on pkg/kapp/diff/resource_with_history.go:
```diff
@@ -218,5 +211,11 @@ func (resourceWithoutHistory) removeAppliedResAnnKeysMods() []ctlres.ResourceMod
 		},
+		// metadata.managedFields technically is not kapp history, but we want to strip it in same situations we are
+		// stripping kapp history, it is server side apply history after all
+		ctlres.FieldRemoveMod{
```
without this, was this breaking anything? (if so, just for ssa, or for non-ssa operation?)

redbaron replied:
Yes, with SSA enabled it was reporting diff on `managedFields` where it shouldn't

cppforlife commented on pkg/kapp/resources/mod_string_map_append.go:
```diff
@@ -29,6 +31,17 @@ func (t StringMapAppendMod) Apply(res Resource) error {
+		kvs[k] = v
+	}
+	unstructured.SetNestedField(obj, kvs, t.Path.AsStrings()...)
+	b, _ := json.Marshal(obj)
```
lets add `if err != nil { panic("Internal inconsistency") }`

cppforlife reviewed (commented):
looking pretty good. ive left some nit-picks, and a few "designy" changes to hopefully simplify a little bit of code.

redbaron commented:
Thanks for a quick review, I'll make requested changes shortly

redbaron committed:
Smallest nits

redbaron committed:
Localize force param value calculation

Also rename SSAFlags struct fields

redbaron committed:
Duplicate SSAFlags in SSAOpts and move them back to cmd/tools

redbaron committed:
Remove dead code

redbaron committed:
Don't try to resolve update conflict when using SSA

cppforlife commented:
@redbaron lmk, when to take a look at the PR again.

ebma16 commented:
Hey guys,

Could you give me a brief update on the topic? Are there still plans to support server-side-apply or will there be no further work on this topic for the time being? 
I only ask because in my project I need to update some Kubernetes resources partially shared with Carvel as well as with other tools (e.g. Operator SDK) and the other tools support SSA.

praveenrewar commented:
Hi @ebma16! Unfortunately the maintainers don't have the bandwidth to work on this at the moment, and I am not sure when we will be able to prioritise it. We can definitely prioritise a review if you would be interested in submitting a PR or continuing on this one.

kapp [rebase rules](https://carvel.dev/kapp/docs/v0.62.x/config/#rebaserules) should still work well with other tools that support SSA. Could you share the specific problem that you are facing?

ebma16 commented:
Hi @praveenrewar,
thank's for the update and your quick response! 

At the moment I have no problems with the kapp rebase rules. We are still in an early phase of the project and I just did a small test for the interaction between the Java-Operator-SDK (using SSA) and the kapp rebase rules (without SSA). So far I could not find any problems and everything worked. I was just curious about the current status when I read this thread ;)

I will continue with this constellation (kapp without SSA and other tools using SSA). If I notice any problems, I will definitely let you know and continue this thread!

praveenrewar commented:
Sure, keep us updated. Happy to chat on [carvel](https://kubernetes.slack.com/archives/CH8KCCKA5) slack channel too if you need help with anything.
